e2e/op_stack.go

```diff
@@ -0,0 +1,282 @@
+package e2e
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"math"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-batcher/batcher"
+	"github.com/ethereum-optimism/optimism/op-batcher/compressor"
+	opbatchermetrics "github.com/ethereum-optimism/optimism/op-batcher/metrics"
+	opnodemetrics "github.com/ethereum-optimism/optimism/op-node/metrics"
+	opnode "github.com/ethereum-optimism/optimism/op-node/node"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/driver"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/sync"
+	opproposermetrics "github.com/ethereum-optimism/optimism/op-proposer/metrics"
+	"github.com/ethereum-optimism/optimism/op-proposer/proposer"
+	opcrypto "github.com/ethereum-optimism/optimism/op-service/crypto"
+	"github.com/ethereum-optimism/optimism/op-service/dial"
+	"github.com/ethereum-optimism/optimism/op-service/sources"
+	"github.com/ethereum-optimism/optimism/op-service/txmgr"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/polymerdao/monomer/utils"
+	"github.com/sourcegraph/conc"
+)
+
+type OPEventListener interface {
+	// LogWithPrefix is meant to be used as a general-purpose log function.
+	// The prefix will never be empty and r will never be nil.
+	// It is only used to fulfill log interfaces defined by third-party OP Stack components.
+	LogWithPrefix(prefix string, r *log.Record)
+	AfterStartup()
+	BeforeShutdown()
+}
+
+type OPStack struct {
+	// The http address (without the http://) of the Ethereum EL node.
+	l1Address string
+	// The websocket address (without the ws://) of the execution engine.
+	engineAddress string
+	// The http address (without the http://) of the op-node.
+	nodeAddress         string
+	privKey             *ecdsa.PrivateKey
+	rollupConfig        *rollup.Config
+	l2OutputOracleProxy common.Address
+	eventListener       OPEventListener
+}
+
+// TODO setup verifiers
+
+func NewOPStack(
+	l1Address,
+	engineAddress,
+	opNodeAddress string,
+	l2OutputOracleProxy common.Address,
+	privKey *ecdsa.PrivateKey,
+	rollupConfig *rollup.Config,
+	eventListener OPEventListener,
+) *OPStack {
+	return &OPStack{
+		l1Address:           l1Address,
+		engineAddress:       engineAddress,
+		nodeAddress:         opNodeAddress,
+		privKey:             privKey,
+		rollupConfig:        rollupConfig,
+		l2OutputOracleProxy: l2OutputOracleProxy,
+		eventListener:       eventListener,
+	}
+}
+
+func (op *OPStack) Run(parentCtx context.Context) (err error) {
+	var wg conc.WaitGroup
+	defer wg.Wait()
+	ctx, cancel := context.WithCancelCause(parentCtx)
+	defer func() {
+		op.eventListener.BeforeShutdown()
+		cancel(err)
+		err = utils.Cause(ctx)
```

Comment:
```suggestion
		err = utils.Cause(ctx)
		cancel(err)
```

Reply:
Not sure I follow this change. If `err != nil` and the context hasn't been canceled, won't this cause us to incorrectly return `nil`?